Statistics3-3

Mukhammed003 opened the pull request:
Выбранная командой

архитектура: MVVM;
способ вёрстки: UIKit.
Ссылка на таск трекер: https://github.com/users/zanpakt0/projects/1/views/8

Сам файл .md с декомпозицией: 
[statistics_Nurmukhanov.md](https://github.com/user-attachments/files/24025733/statistics_Nurmukhanov.md)

Первоначальная оценка времени выполнения задач и реальное время, потраченное на решении каждой задачи:

## Модуль 3:
#### Верстка экрана коллекции NFT
- Настройка navigationBar для кнопки назад (est: 1 час; fact: 20 минут).
- Сверстать коллекцию (est: 2 часа; fact: 1 час 27 минут).
- Сверстать ячейку коллекции (est: 3 часа; fact: 2 часа 56 минут).
- Сверстать Alert об ошибке загрузки данных (est: 1 час; fact: 33 минуты).


#### Логика экрана коллекции NFT
- Логика индикатора загрузки коллекции (est: 1 час; fact: 1 час).
- Подстановка данных в каждую ячейку коллекции (est: 2 часа; fact: 14 минут).
- Настроить взаимосвязи между model, view и view model для паттерна MVVM (est: 2 часа; fact: 25 минут).
- Дополнение структур моделей (est: 1 час; fact: 31 минута).


#### Работа с сетевыми данными
- PUT запросы для добавления nft в корзину и избранное (est: 2 часа; fact: 2 часа 3 минуты).
- GET запрос картинки, имени, стоимости и рейтинга NFT (est: 2 часа; fact: 1 час).
- GET запрос для получения списка понравившихся nft, и списка nft в корзине (est: 2 часа; fact: 1 час 27 минут).

Mukhammed003 committed:
develope(newFeature): Дополнена структура моделей

Выполнены следующие работы:
1) Добавлена структура NftCellViewData для показа данных в коллекции Nft;
2) Добавлена структура NFT для декодирования ответа от сервера на запрос: Get NFT by id.

Mukhammed003 committed:
develope(newFeature): Настроена взаимосвязь между model, view и view model для паттерна MVVM

Выполнены следующие работы:
1) Обновлён UserCollectionViewModel добавлением нужных переменных для связки с view(UserCollectionViewController);
2) Обновлён UserCollectionViewController добавлением метода bindViewModel для связки с view model (UserCollectionViewModel).

Mukhammed003 committed:
develope(newFeature): Добавлено взаимодействие с сетью по получению данных о nft

Выполнены следующие работы:
1) Добавлена структура GetNftRequest для конфигурации запроса по получению данных о nft;
2) Обновлён UserCollectionViewModel с добавлением метода fetchNfts по загрузке nfts;
3) Обновлён NftService с добавлением метода:
         func getNft(id: String, completion: @escaping (Result<NFT, Error>) -> Void) -> NetworkTask? {}.

Mukhammed003 committed:
develope(newFeature): Свёрстан Alert об ошибке загрузки данных

Выполнены следующие работы:
1) Добавлен Extensions файл с реализацией метода universalErrorAlert который отображает алерт об ошибке загрузки данных;
2) Обновлены вью контроллеры(StatisticsViewController, UserCardViewController, UserCollectionViewController) с добавлением метода universalErrorAlert для показа алерта об ошибке загрузки данных;
3) Обновлён NftSerivce, в методе getNft используется кастомная структура GetNftRequest.

Mukhammed003 committed:
develope(newFeature): Разработана логика показа индикатора загрузки коллекции

Выполнены следующие работы:
1) Обновлён Colors с добавлением цвета для фона индикатора загрузки;
2) Общие enum перенесены в Constants;
3) Обновлён Extensions добавлением метода(universalOpenPage) для анимированного открытия новой страницы;
4) Обновлён UserCollectionViewController с добавлением индикатора загрузки.

Mukhammed003 committed:
develope(newFeature): Свёрстана кастомная ячейка для коллекции nft

Выполнены следующие работы:
1) Реализован UserCollectionViewCell с методами для конфигурации ячейки(сам метод для конфигурации не реализован);
2) Обновлён UserCollectionViewModel с добавлением нужных методов и параметров для будущего взаимодействия с сетью (для запросов по получению списка понравившихся (nft) и списка nft добавленные в корзину);
3) Добавлены необходимые иконки(в assets) для отображения кнопки "Добавить в корзину/Убрать из корзины".

Mukhammed003 committed:
develope(newFeature): Свёрстана коллекция для nft

Выполнены следующие работы:
1) Обновлены assets с добавлением корректных картинок для иконки корзины;
2) Обновлён UserCollectionViewCell с добавлением метода configure для установки данных в ячейке;
3) Обновлён UserCollectionViewController с добавлением методов по конфигурации ячеек, и с добавлением методов по установке размеров для коллекции и его ячеек;
4) Обновлён UserCollectionViewModel с добавлением метода prepareNftCellData для подготовки данных для коллекции;
5) Обновлены структуры для NftData(для коллекции nft).

Mukhammed003 committed:
develope(newFeature): Реализован GET запрос для получения списка понравившихся nft, и списка nft в корзине

Выполнены следующие работы:
1) Добавлены структуры: Order, ListOfOrderedNftsData, GetOrderedNftsRequest для работы с запросом для получения списка nft в корзине;
2) Добавлены структуры: Profile, ListOfFavouriteNftsData, GetFavouritesNftsRequest для работы с запросом для получения списка понравившихся nft;
3) Обновлён UserCollectionViewModel с добавлением метода loadEverything для последовательной подгрузки всех данных;
4) Обновлён UserCollectionViewModel с добавлением методов fetchFavouritesNftsList, fetchOrderedNftsList для подгрузки списка понравившихся nft и списка nft в корзине;
5) Обновлён NftService с добавлением методов getFavouritesNft, getOrderedNft.

Mukhammed003 committed:
develope(newFeature): Реализованы PUT запросы для добавления в корзину и избранное

Выполнены следующие работы:
1) Обновлён NftService с добавлением методов putToFavoritesNft, putToOrderedNft для отправления запросов по добавлению nft в корзину и избранное;
2) Добавлены структуры: PutFavouritesNftsRequest, PutOrderedNftsRequest для работы с запросоми по добавлению nft в корзину и избранное;
3) Обновлён UserCollectionViewController с добавлением метода showErrorAlertWhenTappingButtonsInCell для показа алерта об ошибке во время лайка nft или добавления nft в корзину.

Mukhammed003 committed:
fix(bugs): Последние штрихи для читаемости.

Mukhammed003 committed:
develope(newFeature): Обновил файл с декомпозицией эпика

d00m1r commented on FakeNFT/DesignSystem/Colors.swift:
```diff
@@ -66,15 +69,48 @@ extension UIColor {
     }
+    
+    static let forActivityIndicatorBackground = UIColor { traits in
+        switch traits.userInterfaceStyle {
```
В данном случае, ключевое слово `return` везде можно опустить

d00m1r commented on FakeNFT/Scenes /Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -309,33 +297,17 @@ final class UserCardViewController: UIViewController, LoadingView, ErrorView {
-        
-        let retryAction = UIAlertAction(title: errorModel.actionText, style: .default) {_ in
-            errorModel.action()
+        self.universalErrorAlert {[weak self] in
```
Лучше излишне не указывать обращения к `self.`

d00m1r replied:
Принято ставить 1 пробел преред `[weak self]`

d00m1r commented on FakeNFT/Scenes /Statistics/UserCollection/UserCollectionViewCell.swift:
```diff
@@ -1,5 +1,247 @@
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Можно просто указывать `nil`

Mukhammed003 committed:
fix(bugs): Поправил отступы по замечаниям ревьюера.

Mukhammed003 committed:
fix(bugs): Поправил по замечаниям ревьюера.

Выполнены следующие работы:
1) Опущены лишние return;
2) Убраны лишние self;
3) fatalError заменены nil.
